File: HealthLink_BE/src/main/java/com/HealthLink/dto/auth/RegisterRequest.java
package com.HealthLink.dto.auth;

import jakarta.validation.constraints.*;
import lombok.*;

/**
 * Request đăng ký tài khoản mới.
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
public class RegisterRequest {

    @NotBlank(message = "Username is required")
    @Size(min = 3, max = 50, message = "Username must be between 3 and 50 characters")
    private String username;

    @NotBlank(message = "Email is required")
    @Email(message = "Invalid email format")
    private String email;

    @NotBlank(message = "Password is required")
    @Size(min = 6, message = "Password must be at least 6 characters")
    private String password;

    @NotBlank(message = "Phone number is required")
    @Pattern(regexp = "^\\d{9,15}$", message = "Invalid phone number format")
    private String phoneNumber;

    /**
     * Role mặc định sẽ là "Patient" nếu không truyền.
     * Các giá trị hợp lệ: "Patient", "Doctor", "Pharmacy".
     */
    private String role;

    // -------------------------------------------------------------------------
    // Các trường tùy chọn dành cho Patient (có thể null nếu không nhập)
    // -------------------------------------------------------------------------

    /** Ngày sinh, ví dụ: "1999-05-15" (ISO 8601) */
    private String dateOfBirth;

    /** Giới tính: "Male", "Female", "Other" */
    private String gender;

    /** Chiều cao (cm) */
    private Double heightCm;

    /** Cân nặng (kg) */
    private Double weightKg;

    /** Ngôn ngữ ưu tiên: "Vietnamese", "English" */
    private String preferredLanguage;
}

File: HealthLink_MB/lib/models/auth/register_request.dart
/// Model cho request đăng ký tài khoản mới.
/// Tất cả field optional (ngoài required) đều nullable – backend sẽ bỏ qua khi null.
class RegisterRequest {
  // Required
  final String username;
  final String email;
  final String password;
  final String phoneNumber;
  final String role;

  // Optional
  final String? dateOfBirth;
  final String? gender;
  final double? heightCm;
  final double? weightKg;
  final String? preferredLanguage;

  const RegisterRequest({
    required this.username,
    required this.email,
    required this.password,
    required this.phoneNumber,
    this.role = 'Patient',
    this.dateOfBirth,
    this.gender,
    this.heightCm,
    this.weightKg,
    this.preferredLanguage,
  });

  /// Chuyển sang JSON để gửi lên API.
  /// Loại bỏ key null để tránh gửi trường không cần thiết.
  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'username':    username,
      'email':       email,
      'password':    password,
      'phoneNumber': phoneNumber,
      'role':        role,
    };
    if (dateOfBirth      != null) map['dateOfBirth']      = dateOfBirth;
    if (gender           != null) map['gender']           = gender;
    if (heightCm         != null) map['heightCm']         = heightCm;
    if (weightKg         != null) map['weightKg']         = weightKg;
    if (preferredLanguage != null) map['preferredLanguage'] = preferredLanguage;
    return map;
  }
}
